students/ChrisH/lesson09/mailroom.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# -----------------------------------------------------------
+# mailroom.py
+#  Script to automate writing thank you emails to donors.
+# -----------------------------------------------------------
+
+import time
+from sys import exit
+import pickle
+
+
+class Donors(object):
+
+    DATA_FILE = 'donors.pkl'
+
+    def __init__(self):
+        self.donorlist = []
+
+    def add_donor(self, donor):
+        self.donorlist.append(donor)
+
+    def print_report(self):
+        """
+        Prints a formatted report on the donors with name, amount given, number of gifts, and average gift.
+        :return: None
+        """
+        # Find longest name in donor list, or use name_min value
+        name_max = max(*[len(donor.name) for donor in self.donorlist], 25)
+
+        rpt_title = "Donor Name" + ' ' * (name_max - 9) + "| Total Given | Num Gifts | Average Gift"
+        print(rpt_title)
+        print("-" * len(rpt_title))
+        for donor in self.donorlist:
+            print(f"{donor.name:{name_max}}  ", end='')
+            ddons = donor.donations
+            print(f"$ {sum(ddons):>10.2f}   {len(ddons):>9}  ${sum(ddons)/len(ddons):>12.2f}")
+
+    def send_letters_all(self):
+        """
+            Runs through donor data structure, generates a thank you letter for each and saves it to
+            the current working directory with in a date+donorname.txt file
+            :return: None
+            """
+        for donor in self.donorlist:
+            print(f'Generating letter for {donor.name}')
+            now = time.localtime()
+            f_name = f"{now.tm_year}{now.tm_mon:02}{now.tm_mday:02}_"
+            f_name += donor.name.replace(" ", "_") + ".txt"
+            with open(f_name, 'w') as file_out:
+                file_out.write(donor.generate_letter())
```

Comment:
might want to consider `'{:%Y-%m-%d}'.format(datetime.datetime.now())` for your timestamp.